RD-3197 Moved product header to sidebar and removed top bar

kubama opened the pull request:
<!--- Provide JIRA issue ID and a general summary of your changes in the Title above -->

## Description
Moved product header to sidebar and removed top bar. Includes two minor general sidebar related fixes:
- fixed system menu items shrinking on small vertical resolution
- fixed embedded widgets positioning
<!--- Describe your changes to help reviewers understand the details. -->

## Screenshots / Videos
Collapsed:
![image](https://user-images.githubusercontent.com/49753069/144389612-9fe854e0-d42a-4108-ab9f-e508275173ed.png)

Expanded:
![image](https://user-images.githubusercontent.com/49753069/144389686-6bbf86ce-8a85-477e-8ebb-97abc89dfad0.png)

<!--- If applicable, add screenshot(s) or video(s) describing the changes you made. -->
<!--- Consider adding comparison screenshot(s) - before and after the change. -->
<!--- If not applicable, just write “N/A” in this section. -->

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [X] I verified if that change requires a documentation update.
- [X] I added proper labels to this PR.

## Tests
https://jenkins.cloudify.co/view/UI/job/Stage-UI-System-Test/1614/
<!--- If applicable, describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If applicable, provide a link to system tests, e.g: -->
<!--- [#1039](https://jenkins.cloudify.co/.../1039) (2021-08-06 07:05:50) --> 
<!--- If not applicable, just write “N/A” in this section. -->

## Documentation
Will be covered in separate ticket
<!--- If applicable, describe how you documented or plan to document your changes. -->
<!--- Provide links to JIRA issues, other PRs or related documentation pages. -->
<!--- If not applicable, just write “N/A” in this section. -->

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,94 @@
+            left: -23,
+            width: 55,
+            textTransform: 'uppercase',
+            fontFamily: 'JosefinSans-Bold',
```
Referencing the font family that way has some flaws:
1. It depends on `LogoLabel` component where that font is actually included. When we'll remove `LogoLabel` component then this font will become unavailable in the package.
2. White-labelling might be difficult (as customers will need to override local styles)

I'm wondering how difficult it would be to make kind of wrapper function, e.g. `getProductFont` which will get all the necessary details - import CSS file, take into account white-labelling and return font name? We could use it in all places where "JosefinSans-Bold" is referenced directly.

kubama replied:
> It depends on LogoLabel component where that font is actually included. When we'll remove LogoLabel component then this font will become unavailable in the package.

I moved the font declaration CSS import to top level.

> White-labelling might be difficult (as customers will need to override local styles)

The only minor difficulty is that overriding CSS might need to use `!important` to override local style. But as we are using inline styles everywhere then this is a general thing.
IMO more important issue in terms of white-labelling is that we are not assigning class names or ids that could be referenced in overriding CSS.

> I'm wondering how difficult it would be to make kind of wrapper function, e.g. getProductFont which will get all the necessary details - import CSS file, take into account white-labelling and return font name? We could use it in all places where "JosefinSans-Bold" is referenced directly.

I'm not sure what you mean by "taking white-labelling into account".
It could be possible to define font name as SCSS variable and then import it, but the benefit would be quite minor and it would need another cloudify-ui-common release, so I think it's a bit out of scope for this PR.

qooban replied:
> I moved the font declaration CSS import to top level.

I'm not sure about moving it to top level. Maybe we just need to reuse `LogoLabel`? (see my comments in the latest review)

> IMO more important issue in terms of white-labelling is that we are not assigning class names or ids that could be referenced in overriding CSS.

OK, so then if with class names we'll be able to style product header, then let's add them to relevant product header items.

> I'm not sure what you mean by "taking white-labelling into account".
> It could be possible to define font name as SCSS variable and then import it, but the benefit would be quite minor and it would need another cloudify-ui-common release, so I think it's a bit out of scope for this PR.

I didn't have final solution in mind, but right now when I'm thinking about it: maybe we can just add product font name into white-labelling theme `conf/userConfig.json` and use it from there? I'm not sure it is the right solution, I'm looking for the best solution taking into account ease of maintenance and customization flexibility.

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,94 @@
+    const productVersion = useSelector((state: ReduxState) => state.manager.version.version);
+
+    return (
+        <div style={{ overflow: 'hidden', whiteSpace: 'nowrap', flexShrink: 0 }}>
+            <Logo url={logo} style={{ top: 6, left: licenseLabelKey ? 3 : -3 }} />
+            <span
+                style={{
+                    fontFamily: 'JosefinSans-Bold',
+                    textTransform: 'uppercase',
+                    letterSpacing: '0.2em',
+                    marginRight: '1em'
+                }}
+            >
+                {i18n.t('productName')}
+            </span>
+            v<ProductVersion version={productVersion} style={{ color: 'inherit', marginLeft: '-0.3em' }} />
+            {licenseLabelKey && <LicenseLabelWrapper content={t(licenseLabelKey)} linked={!isCommunity} />}
+        </div>
```
General:

Having it all hardcoded using `style` prop might be hard for white-labelling.
I don't have strict suggestion here as we still don't have clear requirements for white-labelling, but recently WR contacted mgmt asking for more flexible solution for customization (one of the suggestions was to use CSS classes or variables for styling, instead of styled-components and React `style` prop).

kubama replied:
If required, we can assign class names to allow custom styling via external CSS file and still use styled-components or style prop.

qooban commented on app/components/sidebar/UserInitialsIcon.tsx:
```diff
@@ -8,9 +8,9 @@ import { ReduxState } from '../../reducers';
 const StyledLabel = styled(Label)`
     &&&&& {
         float: none;
-        margin-left: -10px;
```
I see that you did similar changes in `UserInitialsIcon` and `SideBarItemIcon`.

Is it possible somehow for `UserInitialsIcon` and `SideBarItemIcon` to share common styles to avoid synchronization issues in the future?

kubama replied:
Done

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,94 @@
+    </div>
+);
+
+const LicenseLabelWrapper = ({ content, linked }: { content: string; linked: boolean }) => {
```
I know that `LicenseLabelWrapper` is only used in `SideBarHeader`, but IMHO for clarity it seems to me reasonable to have it in separate file (L12 - L52 are about half of this file). WDYT?

qooban replied:
It seems that the code was copied from `LicenseTag`. I understand that both are required (because  About page or License Management page are not changed yet), so I suggest to extract common parts from both components first and have two version of that tag component.

kubama replied:
Addressed both comments

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,94 @@
+    );
+    const isTrial = useSelector((state: ReduxState) => state.manager.license.data?.trial);
+
+    let licenseLabelKey;
+    if (isCommunity) {
+        licenseLabelKey = 'community';
+    } else if (isExpired) {
+        licenseLabelKey = 'expired';
+    } else if (isTrial) {
+        licenseLabelKey = 'trial';
+    }
```
Maybe that is something we can wrap using `useMemo`?

kubama replied:
Done

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,94 @@
+    return (
+        <div style={{ overflow: 'hidden', whiteSpace: 'nowrap', flexShrink: 0 }}>
+            <Logo url={logo} style={{ top: 6, left: licenseLabelKey ? 3 : -3 }} />
+            <span
```
Optional:

I think we can make product name a bit bigger and maybe drop letter spacing. WDYT?
As there's separate issue for styling the sidebar I'm adding it as optional.

kubama replied:
I'd leave it for styling ticket

qooban replied:
OK.

qooban reviewed (commented):
Looks good in general 👍🏻 

Provided some suggestions in comments.

Do you have the same rendering issue with license tag when expanding/collapsing the sidebar?

https://user-images.githubusercontent.com/5202105/144407228-92795183-d5ca-4d23-9f63-0c7e112aaabb.mp4

kubama commented:
> Do you have the same rendering issue with license tag when expanding/collapsing the sidebar?

Yes, unfortunately I have no clue what's the reason behind.

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,48 @@
+                        }}
+                    />
+                </div>
+                <Logo url={logo} style={{ position: null, height: 33, margin: '11px 8px' }} />
```
I didn't spot it before - I'm not sure if passing logo directly is the right thing to do. We will lose white-labelling option here. Please take a look at how `Logo` component works in relation to theming.

Maybe we should introduce new parameter in the theme (in `conf/userConfig.json` - `whiteLabel` section) and/or enhance `Logo` component to get another prop indicating which logo version should be used (for white/dark background)? Or maybe just changing the logo in `cloudify-ui-components` will be sufficient? You need to check if there are any other places using the current logo. I think that Blueprints and Blueprint Catalog widgets are using logo, but it looks like it is part of the widget.

kubama replied:
> I didn't spot it before - I'm not sure if passing logo directly is the right thing to do. We will lose white-labelling option here. Please take a look at how Logo component works in relation to theming.
Maybe we should introduce new parameter in the theme (in conf/userConfig.json - whiteLabel section) and/or enhance Logo component to get another prop indicating which logo version should be used (for white/dark background)? Or maybe just changing the logo in cloudify-ui-components will be sufficient? You need to check if there are any other places using the current logo. I think that Blueprints and Blueprint Catalog widgets are using logo, but it looks like it is part of the widget.

My understanding is that current state with two versions of the logo in use is temporary - IIUC eventually we are going to switch to dark background in other places and only one version of the logo will be used. If so, is it worth to introduce new theme parameter at this point? Wouldn't `theme.logo ?? builtInDarkBackgroundLogo` be sufficient?

qooban replied:
OK, as we discussed let's go with `theme.logo ?? builtInDarkBackgroundLogo`.

qooban commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,48 @@
+                    />
+                </div>
+                <Logo url={logo} style={{ position: null, height: 33, margin: '11px 8px' }} />
+                <span
+                    style={{
+                        letterSpacing: '0.2em',
+                        marginRight: '1em',
+                        textTransform: 'uppercase'
+                    }}
+                >
+                    {i18n.t('productName')}
+                </span>
```
I think you should be able to reuse `LogoLabel` component here (I did it locally slightly extending `LogoLabel` customizability by introducing `style` prop there), then we have only one place with `JosefinSans-Bold` font defined, so we can move CSS import there. WDYT?

My example (not necessarily perfect):
```
<LogoLabel
    color="inherit"
    style={{
        fontSize: '18px',
        marginRight: '5px',
        textTransform: 'uppercase',
        display: 'inline'
    }}
    content={i18n.t('productName')}
/>
```
and in `LogoLabel` I just added `style: CSSProperties` to props.

kubama replied:
> I think you should be able to reuse LogoLabel component here (I did it locally slightly extending LogoLabel customizability by introducing style prop there), then we have only one place with JosefinSans-Bold font defined, so we can move CSS import there. WDYT?

As stated in another comment I'm intentionally using `JosefinSans-Bold` font in another place.
Also, in the example that you provided you are overriding all `LogoLabel` style props (except font family) and additionally setting `display` prop to override internal `h2` styling. I don't think it is worth it - it's more complicated than simply using a bare `span`.

qooban replied:
OK, let's keep `span` here for now as from the implementation perspective it's not a big gain.

In the future it could probably be improved when we'll have clear overview what are the expected theme changes (e.g. for About Modal and all the so-called logo-pages).

qooban reviewed (commented):
Added some more comments. 

Mostly related to white-labelling which seems a bit tricky. Also - as per my understanding product header is the most common part of customization - let's try to make the customization as easy as possible with the tools we have right now.

If you'll see that it'll be reasonable to handle changes I'm suggesting in a separate ticket, let's create one and try to fit it in this or next sprint.

kubama commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,48 @@
+
+    return (
+        <div style={{ overflow: 'hidden', whiteSpace: 'nowrap', flexShrink: 0 }}>
+            <span style={{ fontFamily: 'JosefinSans-Bold' }}>
```
After experimenting it turned out this font just looks better in such small size than the main font, so I would keep it.

qooban replied:
OK, as we discussed let's keep it.

kubama commented:
@qooban
My general question after reading your comments is - what's the actual target for customization level? On one hand you are suggesting moving to well defined class names to allow easy CSS overriding, on the other hand you are suggesting extending support for userConfig.json. These two directions are kind of overlapping. If we are going to maintain class names then this will give customers maximum level of flexibility. If this is the case, then is it worth it to spend time on extending userConfig.json support? Is the actual goal for white-labelling well defined at the moment? I think it should be if we are making implementation decisions.

kubama commented:
I addressed code review comments as agreed during the call.
Created https://github.com/cloudify-cosmo/docs.getcloudify.org/pull/2047 to satisfy doc check.

qooban commented on test/cypress/integration/templateManagement/templates_spec.ts:
```diff
@@ -81,7 +81,7 @@ describe('Templates segment', () => {
     });
 
-    it('allows users to create and modify templates', () => {
+    it.only('allows users to create and modify templates', () => {
```
Probably something to change before merge 🙂

kubama replied:
Done

qooban reviewed (commented):
Just one small thing to fix before merging.

NorbertAmartus commented on app/components/LicenseLabel.tsx:
```diff
@@ -0,0 +1,57 @@
+            style={{
+                backgroundColor: '#FFC304',
+                color: '#000000',
+                ...style
```
As I can see in the [Label typing](https://github.com/Semantic-Org/Semantic-UI-React/blob/61a08733941fb60d6ff09f30c3e2fc13e40448da/src/elements/Label/Label.d.ts#L37) it can be given a `className`

Because of that I think that we should use `styled-components` , so that:
- The `styled` prop would be destructurized with other props (component would be simplier)
- We wouldn't pass a plain object into the `style` prop, which would result in a better performance

Example of styled component:
```typescript
const StyledLabel = styled(Label)`
  backgroundColor: '#FFC304',
  color: '#000000'
`;
```

NorbertAmartus replied:
As commented [here](https://github.com/cloudify-cosmo/cloudify-stage/pull/1689#issuecomment-986705037), I'm resolving this issue

NorbertAmartus commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,62 @@
+                >
+                    {i18n.t('productName')}
+                </span>
+                v<ProductVersion version={productVersion} style={{ color: 'inherit', marginLeft: '-0.3em' }} />
```
I think that this `v`, at the beggining of the line, might be a typo 😁

kubama replied:
Actually it's there by design 🙂. It's a prefix before the version number, for example 'v6.3'.

NorbertAmartus replied:
Okay, thanks for explaining :)

NorbertAmartus commented on app/components/sidebar/SideBarHeader.tsx:
```diff
@@ -0,0 +1,62 @@
+            </StyledLink>
+        </div>
+    );
+};
```
Every element (inside this component) is using `style` prop for styling purposes.
I think that it would be more advised to use `styled-components`, so that:
- Whole component would be more performant (simple rerender would add a cost of recreating all of those plain objects)
- Name of the newly created `styled-components` components would be useful in terms of readability of the component (to know what is the purpose of 2 divs at the top of the return statement, I'd have to go through the rest of the `return` declaration, to understand that)

Let me know what do you think

NorbertAmartus replied:
As commented [here](https://github.com/cloudify-cosmo/cloudify-stage/pull/1689#issuecomment-986705037), I'm resolving this issue

NorbertAmartus commented on app/components/sidebar/UserInitialsIcon.tsx:
```diff
@@ -4,22 +4,23 @@ import { useSelector } from 'react-redux';
-    return <StyledLabel circular>{username.substr(0, 2).toUpperCase()}</StyledLabel>;
+    return (
+        <StyledLabel style={defaultStyle} circular>
+            {username.substr(0, 2).toUpperCase()}
```
We could extract that to another, memoized variable, so that it would be explicitly declared what is the purpose of that line

Without that variable, me as an another developer, might look into the application (or guess by the file name), to understand why this line is used like it is 😁

I think that it should be explicitly said, so that we wouldn't have to guess.

Implementation example:

```typescript
const userInitials = useMemo(() => {
  return username.substr(0, 2).toUpperCase();
}, [username]);

return (
  <StyledLabel style={defaultStyle} circular>
    {userInitials}
  <StyledLabel />
)
```

kubama replied:
Done

NorbertAmartus commented on app/components/templateManagement/pages/PageManagement.tsx:
```diff
@@ -215,7 +223,7 @@ export default function PageManagement({ pageId, isEditMode = false }: PageManag
         <div className="main">
             <div className="sidebarContainer">
-                <Sidebar visible as={Menu} vertical size="small">
+                <Sidebar visible as={Menu} vertical size="small" style={{ width: expandedSidebarWidth }}>
```
Sidebar component supports `className` property
Because of that I think that we may favour the declaration which may be using `styled-components` (for performance and readability reasons, which was mentioned in the previous comments)

Implementation example:
```typescript
const StyledSidebar = styled(Sidebar)`
  width: ${expandedSidebarWidth}
`
```

NorbertAmartus replied:
As commented [here](https://github.com/cloudify-cosmo/cloudify-stage/pull/1689#issuecomment-986705037), I'm resolving this issue

NorbertAmartus reviewed (commented):
Looks good 💪
Just added some comments mostly related to usage of the inline `style` prop

kubama commented:
@NorbertAmartus

Regarding your styling comments - at the moment our established approach is that inline styling as well as styled-components are equally good approaches, please see https://cloudifysource.atlassian.net/wiki/spaces/RD/pages/435388492/Cloudify+UI+Code+Style#Rules. If you think it is something worth changing and that we should prefer one approach over the other then lets discuss your points on our weekly.

NorbertAmartus commented:
> @NorbertAmartus
> 
> Regarding your styling comments - at the moment our established approach is that inline styling as well as styled-components are equally good approaches, please see https://cloudifysource.atlassian.net/wiki/spaces/RD/pages/435388492/Cloudify+UI+Code+Style#Rules. If you think it is something worth changing and that we should prefer one approach over the other then lets discuss your points on our weekly.

Okay, sure
In that case I'm resolving the issues related to the `style` prop
